mkdir_linux.go

```diff
@@ -0,0 +1,440 @@
+	// drop the link from the stack.
+	linkUnwalked []string
+}
+
+func (se symlinkStackEntry) String() string {
+	return fmt.Sprintf("<%s>/%s [->%s]", se.dir.Name(), se.remainingPath, strings.Join(se.linkUnwalked, "/"))
+}
+
+func (se symlinkStackEntry) Close() {
+	_ = se.dir.Close()
+}
+
+type symlinkStack struct {
+	linkStack []*symlinkStackEntry
+}
+
+func (s symlinkStack) IsEmpty() bool {
+	return len(s.linkStack) == 0
+}
+
+func (s *symlinkStack) Close() {
+	for _, link := range s.linkStack {
+		link.Close()
+	}
+	// TODO: Switch to clear once we switch to Go 1.21.
+	s.linkStack = nil
+}
+
+func (s *symlinkStack) Push(dir *os.File, remainingPath, linkTarget string) error {
+	// Split the link target and clean up any "" parts.
+	linkTargetParts := slices.DeleteFunc(
+		strings.Split(linkTarget, "/"),
+		func(part string) bool { return part == "" })
+
+	// Don't add a no-op link to the stack. You can't create a no-op link
+	// symlink, but if the symlink is /, partialOpen has already jumped to the
+	// root and so there's nothing more to do.
+	if len(linkTargetParts) == 0 {
+		return nil
+	}
+
+	// Copy the directory so the caller doesn't close our copy.
+	dirCopy, err := dupFile(dir)
+	if err != nil {
+		return err
+	}
+
+	// Add to the stack.
+	s.linkStack = append(s.linkStack, &symlinkStackEntry{
+		dir:           dirCopy,
+		remainingPath: remainingPath,
+		linkUnwalked:  linkTargetParts,
+	})
+	return nil
+}
+
+var errBrokenSymlinkStack = errors.New("[internal error] broken symlink stack")
+
+func (s *symlinkStack) PopPart(part string) error {
+	if s.IsEmpty() {
+		// If there is nothing in the symlink stack, then the part was from the
+		// real path provided by the user, and this is a no-op.
+		return nil
+	}
+	tailEntry := s.linkStack[len(s.linkStack)-1]
+
+	// Double-check that we are popping the component we expect.
+	headPart := tailEntry.linkUnwalked[0]
+	if headPart != part {
+		return fmt.Errorf("%w: trying to pop component %q but the last stack entry is %s (%q)", errBrokenSymlinkStack, part, tailEntry, headPart)
+	}
+
+	// Drop the component, and the entry if that was the last component.
+	tailEntry.linkUnwalked = tailEntry.linkUnwalked[1:]
+	if len(tailEntry.linkUnwalked) == 0 {
+		s.linkStack = s.linkStack[:len(s.linkStack)-1]
+		tailEntry.Close()
+	}
+	return nil
+}
+
+func (s *symlinkStack) PopLastSymlink() (*os.File, string, bool) {
+	if s.IsEmpty() {
+		return nil, "", false
+	}
+	tailEntry := s.linkStack[len(s.linkStack)-1]
+	s.linkStack = s.linkStack[:len(s.linkStack)-1]
+	return tailEntry.dir, tailEntry.remainingPath, true
+}
+
+// partialOpen tries to lookup as much of the request path as possible and
+// opens the final existing component of the requested path, returning a file
+// handle to the final existing component and a string containing the remaining
+// path components (to be later created by MkdirAll).
+func partialOpen(root *os.File, unsafePath string) (_ *os.File, _ string, Err error) {
```

Comment:
Okay, I see what I was missing here: https://github.com/cyphar/filepath-securejoin/pull/13#discussion_r1655098264

This is faking "RESOLVE_IN_ROOT". So links to / are interpreted to links to root. I'd at least document that here.

IMHO, we should test this function too. It is very tricky. But I wouldn't go for that now, I think we can add tests once we see if there is a simpler way to do this. I'd really like a simpler alternative.

Reply:
Yes, this is emulating `RESOLVE_IN_ROOT` (just like `SecureJoin` does, but without a broken API).

I don't really see a simpler alternative to:

1. Look up as much of the `MkdirAll` path, returning `(last-existing-component, remaining-components)`.
2. `mkdir` each of the remaining components.

Yes, the userspace implementation of (1) is complicated -- that's why I implemented `openat2(RESOLVE_IN_ROOT)` :smile_cat:.

I don't see a way you can re-use `SecureJoin` for `partialOpen`, precisely because the semantics of `SecureJoin` are wrong for writing safe programs...

I will admit that the symlink stack logic in `partialOpen` is quite a bit of extra logic that I don't really like, but if we want to have identical semantics to `openat2`, we need to emulate that aspect as well. (And in this case, it's only used to change the return value when we hit a non-existent path. It isn't used to change the control flow of the algorithm.)

> Clearly secureJoin can be implemented by using partialOpen() and in the future we should probably do that. This might be ok for now.

In principle, you could delete most of `SecureJoin` and just use `partialOpen`. For paths that exist, the semantics would be identical. I would love to remove as much as of the duplication between `SecureJoin` and `partialOpen` as possible, don't get me wrong!

However, I didn't do it in this PR because the semantics are noticeably different when dealing with dangling symlinks and non-existent paths, and I am not entirely sure you can get all of the semantics to be identical.

(I would argue one of the main API problems with `SecureJoin` is precisely the fact that you can use it with non-existent paths. There really isn't a reasonable result to give when you ask for a non-existent path.)

I was thinking of looking into this for a possible v2 of `securejoin`, if it makes sense...

Reply:
I've moved `partialOpen` to a separate file and renamed it to `partialLookupInRoot` to make it clearer that this is a generic helper function which could be used to implement `SecureJoin` on Linux (with some caveats).